### PR TITLE
Fix leading spaces pushing text outside Label autowrap boundary in 3.x

### LINKED ISSUE #59985: Leading spaces may push text outside `Label` autowrap boundary in 3.x
### Godot version

3.x (14856b1dbd7b428341f00f8819d53935d555b72c)

### System information

Arch Linux, GLES 3, Mesa Intel(R) UHD Graphics (CML GT2)

### Issue description

When autowrap is enabled, long leading spaces in `Label` may push text outside the boundary.

| Left Align | Center Align | Right Align | Fill |
| -- | -- | -- | -- |
| ksnip_20220407-221947 | ksnip_20220407-222001 | ksnip_20220407-222013 | ksnip_20220407-222024 |

Not in `master`. I believe it has been like this for quite a few versions.

### Steps to reproduce

Create a Label, enable Auto Wrap, put lots of spaces before actual text.

### Minimal reproduction project

[LabelSpace.zip](https://github.com/godotengine/godot/files/8443780/LabelSpace.zip)

### PR DESCRIPTION
**Pull request by ZheisterCoding**
Fixed #59985 leading spaces autowrap boundary problem

Bug problem:
No condition for when the first character of the label is a space character

Fix:
Added an IF condition for when the first character is a space character
The autowrap boundary treat this space character as another dummy word in the WordCache linked list and proceed to function normally

_*Note: the first screenshot is post bug fix, the second screenshot is pre bug fix, the fix works for left, center, right align_

Screenshot of post-bug fix
Screenshot of Pre-bug fix

### CODE REVIEW

**Comment by timothyqiu** [comment]
Texts now go out of the bounary when autowrap is turned off :(

ksnip_20220413-154715

... and it's still pushing text outside the autowrap boundary on the second line:

ksnip_20220413-155842

**Comment by ZheisterCoding** [comment]
> Texts now go out of the bounary when autowrap is turned off :(
> 
> ksnip_20220413-154715
> 
> ... and it's still pushing text outside the autowrap boundary on the second line:
> 
> ksnip_20220413-155842

hmmm i think i must have forgotten to add an else-if statement, let me fix it

**Comment by ZheisterCoding** [comment]
@timothyqiu Hi, I try testing different scenarios similar to yours while toggling di Autowrap checkbox in the video below, it seems to be working fine. How did the problem re-appear?

https://user-images.githubusercontent.com/50128969/163136095-dd457e3c-e4fc-493e-9400-1af1163fdff8.mp4

**Comment by timothyqiu** [comment]
@ZheisterCoding It's already there in your screen record. You can zoom in to get a better view.

**Comment by ZheisterCoding** [comment]
> @ZheisterCoding It's already there in your screen record. You can zoom in to get a better view.

Hi, I noticed the problem as well after you mentioned this. I've made a small adjustment, the problem should disappear now hopefully. @timothyqiu

**Comment by timothyqiu** [comment]
The first issue is fixed. But the second issue (leading spaces on the second line) still exists.

**Comment by ZheisterCoding** [comment]
> The first issue is fixed. But the second issue (leading spaces on the second line) still exists.

@timothyqiu Thank you for pointing that out.  I've added a local boolean variable within the function to track when a newline is made, so the leading space issue for the very first line and created newlines should now hopefully be fixed.

https://user-images.githubusercontent.com/50128969/163146916-8e2aba3b-4eeb-4e24-903b-ede9f3f5b9a7.mp4

File: scene/gui/label.cpp [hunk 1]
```diff
@@ -447,12 +452,20 @@ void Label::regenerate_word_cache() {
...
 			if (i < xl_text.length() && xl_text[i] == ' ') {
-				if (line_width > 0 || last == nullptr || last->char_pos != WordCache::CHAR_WRAPLINE) {
+				if (i == 0 || (i != 0 && prev_insert_newline)) {
+					if (current_word_size == 0) {
+						word_pos = i;
+					}
+					current_word_size += space_width;
+					line_width += space_width;
+				}
+				else if (line_width > 0 || last == nullptr || last->char_pos != WordCache::CHAR_WRAPLINE) {
```
Comment (timothyqiu) [inline scene/gui/label.cpp:468]:
```suggestion
				} else if (line_width > 0 || last == nullptr || last->char_pos != WordCache::CHAR_WRAPLINE) {
```

Small code style issue :)

File: scene/gui/label.cpp [hunk 2]
```diff
@@ -447,12 +452,20 @@ void Label::regenerate_word_cache() {
...
 			if (current == '\n') {
 				insert_newline = true;
+				prev_insert_newline = true;
 			} else if (current != ' ') {
 				total_char_cache++;
 			}
 
 			if (i < xl_text.length() && xl_text[i] == ' ') {
-				if (line_width > 0 || last == nullptr || last->char_pos != WordCache::CHAR_WRAPLINE) {
+				if (i == 0 || (i != 0 && prev_insert_newline)) {
```
Comment (timothyqiu) [inline scene/gui/label.cpp:461]:
```suggestion
				if (i == 0 || prev_insert_newline) {
```

It's redundant here I believe.

Also, could the `i == 0` check be removed here, and make `prev_insert_newline` default to `true` instead?

**Review by timothyqiu — COMMENTED** [review]
This solves the original issue. But I'm not sure if changing `current_word_size` here is OK. Can you explain what `word_pos`, `current_word_size`, and `line_width` do?

File: scene/gui/label.cpp (hunk 2 above)
Comment (ZheisterCoding) [reply scene/gui/label.cpp:461]: I don't think you can remove the i==0 since the very first original issue was caused by the fact that, the case when xl_text[i] when i = 0 (which is the very first character of the string) is equal to the space character, is missing

**Comment by ZheisterCoding** [comment]
The way that this particular function works is that the string is stored in a linked list named `word_cache` So, the words are stored in a list, which in this case is separated by 'spaces‘。

`word_pos`, `current_word_size` are counters to monitor the current word node (the node is called `WordCache` in the cpp file), word_pos takes the position of the first character in the word node, and current_word_size represents the length. In this case, what I have done is created a temporary word node consisting of exactly 1 space character (normally the node would contain letters). Hence, the `current_word_size` is equals to one `space-width`, the `word_pos` represents the index of the first character in the word, so the dummy one 'space' word node is treated like any other nodes in `WordCache` linked list.

`line_width` affects the autowrap functionality. Basically, autowrap is implemented when the `line_width` exceeds the `width` of the label box (hence, `line_width >= width` in line 492) in the code. The `line_width` takes into account `current_word_size` and `space_count`, in which every space is equals to one `space_width` in length.

So in conclusion, I've converted the singular leading space when i= 0 or a newline is made, into a 'dummy' node, and it is treated like another normal string of alphabetical characters. This allows autowrap functionality to work normally when there is a leading space, since previously, the autowrap didn't detect the leading spaces, which causes the words to go over the boundary.
@timothyqiu

File: scene/gui/label.cpp [hunk 3]
```diff
@@ -429,6 +432,8 @@ void Label::regenerate_word_cache() {
 				space_count = 0;
 			} else if ((i == xl_text.length() || current == '\n') && last != nullptr && space_count != 0) {
 				//in case there are trailing white spaces we add a placeholder word cache with just the spaces
+                prev_insert_newline = false;
```
Comment (timothyqiu) [inline scene/gui/label.cpp:435]: The indent of this line is incorrect.

Otherwise looks good. Please squash the commits and force push. See [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase).

**Comment by ZheisterCoding** [comment]
@timothyqiu I've created a new pull request in here #60216
